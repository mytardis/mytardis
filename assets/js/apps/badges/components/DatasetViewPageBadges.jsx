import React, { Fragment, useState } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { css } from '@emotion/core';
import { fetchDatasetData } from './utils/FetchData';
import Spinner from './utils/Spinner';
import DatasetSizeBadge from './DatasetSizeBadge';
import DatasetExperimentCountBadge from './DatasetExperimentCountBadge';
import DatasetDatafileCountBadge from './DatasetDatafileCountBadge';


const DatasetViewPageBadges = ({ datasetID }) => {
  const [datasetData, setDatasetData] = useState('');
  const [isLoading, setIsLoading] = useState(true);
  const spinnerCss = css`
    margin: auto;
    width: 20%;
    float: right;
    color: 9B9B9B;
  `;
  React.useEffect(() => {
    fetchDatasetData(datasetID).then((data) => {
      setDatasetData(data);
      setIsLoading(false);
    });
  }, []);

  return (
    isLoading ? <Spinner override={spinnerCss} />
      : (
        <Fragment>
          <span className="mr-2">
            <DatasetExperimentCountBadge datasetData={datasetData} />
          </span>
          <span className="mr-2">
            <DatasetDatafileCountBadge datasetData={datasetData} />
          </span>
          <span className="mr-2">
            <DatasetSizeBadge datasetData={datasetData} />
          </span>
        </Fragment>
      )
  );
};

DatasetViewPageBadges.propTypes = {
  datasetID: PropTypes.string.isRequired,
};

const elem = document.querySelector('.badges');
let datasetID = null;
if (elem) {
  [, datasetID] = elem.id.split('-');
  ReactDOM.render(
    <DatasetViewPageBadges datasetID={datasetID} />, elem,
  );
}

export default DatasetViewPageBadges;
