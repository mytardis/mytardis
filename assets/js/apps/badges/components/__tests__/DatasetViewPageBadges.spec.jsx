import React from 'react';
import ReactDOM from 'react-dom';
import { act } from '@testing-library/react';
import DatasetViewPageBadges from '../DatasetViewPageBadges';

global.fetch = require('jest-fetch-mock');

const fakeDatasetData = {
  dataset_datafile_count: 5,
  dataset_experiment_count: 3,
  dataset_size: 14216734,
};
let container = null;
beforeEach(() => {
  jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
    json: () => Promise.resolve(fakeDatasetData),
  }));
  container = document.createElement('div');
  container.setAttribute('id', 'experiment-1234');
  container.setAttribute('class', 'badges');
  document.body.appendChild(container);
});
afterEach(() => {
  // cleanup on exiting
  container.remove();
  container = null;
});

describe('renders badges on experiment view page', () => {
  it('should render all badges', async () => {
    await act(async () => {
      ReactDOM.render(<DatasetViewPageBadges datasetID="123" />, container);
    });
    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('span').length)
      .toEqual(6);
  });
  it('should render experiment count badge', async () => {
    await act(async () => {
      ReactDOM.render(<DatasetViewPageBadges datasetID="123" />, container);
    });
    expect(container.querySelectorAll('span')[0].textContent.trim())
      .toEqual('3');
  });
  it('should render datafile count badge', async () => {
    await act(async () => {
      ReactDOM.render(<DatasetViewPageBadges datasetID="123" />, container);
    });
    expect(container.querySelectorAll('span')[2].textContent.trim())
      .toEqual('5');
  });
  it('should render last updated badge', async () => {
    await act(async () => {
      ReactDOM.render(<DatasetViewPageBadges datasetID="123" />, container);
    });
    expect(container.querySelectorAll('span')[4].textContent.trim())
      .toEqual('13.56 MB');
  });
});
