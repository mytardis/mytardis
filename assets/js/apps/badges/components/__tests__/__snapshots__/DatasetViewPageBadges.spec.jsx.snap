// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders badges on experiment view page should render all badges 1`] = `
<div
  class="badges"
  id="experiment-1234"
>
  <span
    class="mr-2"
  >
    <span
      class="badge badge-info"
      title="In 3 experiments"
    >
      <i
        class="fa fa-cogs"
      />
       
      3
    </span>
  </span>
  <span
    class="mr-2"
  >
    <span
      class="badge badge-info"
      title="5 files"
    >
      <i
        class="fa fa-file"
      />
       
      5
    </span>
  </span>
  <span
    class="mr-2"
  >
    <span
      class="badge badge-info"
      title="Dataset size is"
    >
      13.56 MB
    </span>
  </span>
</div>
`;
